File: JS/workbook2/omok.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("dev/stdin").toString().trim().split("\n")
    : `0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 2 2 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 2 2 2 2 2 2
0 0 0 0 0 0 0 0 0 0 0 0 0 2 2 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 2 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 2 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 2 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 2 2 2 2`.split("\n");

const visited = Array.from(Array(4), () =>
  Array.from(Array(19), () => Array(19).fill(-1))
);

const map = input.map((v, x) =>
  v.split(" ").map((value, y) => {
    if (value === "1") {
      for (let i = 0; i < 4; i++) {
        visited[i][x][y] = 1;
      }
      return "b";
    } else if (value === "2") {
      for (let i = 0; i < 4; i++) {
        visited[i][x][y] = 1;
      }
      return "w";
    } else {
      return value;
    }
  })
);

const dx = [1, 1, 0, -1];
const dy = [0, 1, 1, 1];

let queue = [];

const bfs = () => {
  while (queue.length) {
    const [dir, x, y] = queue.shift();
    const nx = x + dx[dir];
    const ny = y + dy[dir];

    if (nx < 0 || 19 <= nx || ny < 0 || 19 <= ny) {
      if (visited[dir][x][y] === 5) {
        queue = [];
        return true;
      }
      continue;
    }

    if (map[nx][ny] !== map[x][y]) {
      if (visited[dir][x][y] === 5) {
        queue = [];
        return true;
      }
      continue;
    }

    visited[dir][nx][ny] = visited[dir][x][y] + 1;

    queue.push([dir, nx, ny]);
  }
};

for (let i = 0; i < 19; i++) {
  for (let j = 0; j < 19; j++) {
    if (map[j][i] !== "n") {
      for (let k = 0; k < 4; k++) {
        if (visited[k][j][i] === 1) {
          queue.push([k, j, i]);
          const result = bfs();

          if (result) {
            map[j][i] === "b" ? console.log(1) : console.log(2);
            console.log(j + 1, i + 1);
            return;
          }
        }
      }
    }
  }
}

console.log(0);

File: src/workbook2/Omok.java
package workbook2;

import java.io.*;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Omok {
    static int[][][] visited = new int[4][19][19];
    static String[][] map = new String[19][19];
    static int[] dx = {1, 1, 0, -1}, dy = {0, 1, 1, 1};
    static Queue<int[]> queue = new LinkedList<>();

    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 19; j++) {
                Arrays.fill(visited[i][j], -1);
            }
        }

        for (int i = 0; i < 19; i++) {
            String[] s = br.readLine().split(" ");
            for (int j = 0; j < 19; j++) {
                if(s[j].equals("1")){
                    for (int k = 0; k < 4; k++) {
                        visited[k][i][j] = 1;
                    }

                    map[i][j] = "b";
                }else if(s[j].equals("2")){
                    for (int k = 0; k < 4; k++) {
                        visited[k][i][j] = 1;
                    }

                    map[i][j] = "w";
                }else{
                    map[i][j] = "n";
                }
            }
        }

        boolean isEnd = false;
        for (int i = 0; i < 19; i++) {
            for (int j = 0; j < 19; j++) {
                if(!map[j][i].equals("n")){
                    for (int k = 0; k < 4; k++) {
                        if(visited[k][j][i] == 1){
                            queue.add(new int[]{k, j, i});

                            if(bfs()){
                                isEnd = true;
                                if (map[j][i].equals("b")) {
                                    bw.write("1\n");
                                } else {
                                    bw.write("2\n");
                                }
                                bw.write(String.valueOf(j + 1) + " " + String.valueOf(i + 1));
                            }
                        }
                    }
                }
            }
        }

        if(!isEnd) bw.write("0");

        bw.flush();
        bw.close();
    }

    static boolean bfs(){
        while(!queue.isEmpty()){
            int[] xyz = queue.remove();
            int x = xyz[1], y = xyz[2], dir = xyz[0];
            int nx = x + dx[dir], ny = y + dy[dir];

            if(nx < 0 || 19 <= nx || ny < 0 || 19 <= ny){
                if(visited[dir][x][y] == 5){
                    queue = new LinkedList<>();
                    return true;
                }
                continue;
            }

            if(map[nx][ny] != map[x][y]){
                if(visited[dir][x][y] == 5){
                    queue = new LinkedList<>();
                    return true;
                }
                continue;
            }

            visited[dir][nx][ny] = visited[dir][x][y] + 1;
            queue.add(new int[]{dir, nx, ny});
        }

        return false;
    }
}
